### PR TITLE
oauth2: support PKCE

### LINKED ISSUE #603: Feature request: PKCE
Many OAuth providers no longer support many flows without PKCE, yet the `oauth2` library doesn't have any built-in support. Generating code code verifiers and challenges must be done with a third party library, both need to be attached via unspecified Auth URL Parameters, etc.

Would there be interest in merging a PR that provided
- Functions to generate code verifiers and challenges
- Included them in authentication and token exchange flows as part of standard data structures with clear PKCE semantics
- Included working example code in documentation for above

### PR DESCRIPTION
**Pull request by hickford**
Fixes #603

Fixes golang/go#59835

### CODE REVIEW

**Comment by J0** [comment]
Hey 

I'm a user of this package and just wanted to say that better PKCE support would be a great plus for me. Thanks for putting this together and hope this goes through!

File: oauth2.go [hunk 1]
```diff
@@ -208,24 +210,28 @@ func (c *Config) PasswordCredentialsToken(ctx context.Context, username, passwor
...
 	}
 	return retrieveToken(ctx, c, v)
 }
 
-// Client returns an HTTP client using the provided token.
+func (c *Config) Exchange(ctx context.Context, code string, opts ...AuthCodeOption) (*Token, error) {
+	return c.ExchangeWithPKCE(ctx, code, nil, opts...)
+}
+
+`345432Q// Client returns an HTTP client using the provided token.
```
Comment (jkroepke) [inline oauth2.go:234]:
```suggestion
// Client returns an HTTP client using the provided token.
```
seems like type into a wrong window focus

Comment (hickford) [reply oauth2.go]: Thanks fixed

File: pkce.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	// "RECOMMENDED that the output of a suitable random number generator be used to create a 32-octet
+	// sequence.  The octet sequence is then base64url-encoded to produce a 43-octet URL-safe string to use as the code verifier."
+	// https://datatracker.ietf.org/doc/html/rfc7636#section-4.1
+	data := make([]byte, 32)
+	if _, err := io.ReadFull(rand.Reader, data); err != nil {
+		panic(err)
+	}
+	verifier := base64.URLEncoding.EncodeToString(data)
+	sha := sha256.Sum256([]byte(verifier))
+	challenge := base64.URLEncoding.EncodeToString(sha[:])
```
Comment (wzy9607) [inline pkce.go:58]: RFC requires the no padding version of base64 url encoding.

https://datatracker.ietf.org/doc/html/rfc7636#section-3

**Review by wzy9607 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**